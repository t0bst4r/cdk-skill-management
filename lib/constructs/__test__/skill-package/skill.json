{
  "manifest": {
    "manifestVersion": "1.0",
    "publishingInformation": {
      "distributionCountries": [],
      "locales": {
        "de-DE": {
          "description": "My Personal Smart Home",
          "examplePhrases": [],
          "keywords": [
            "Smart Home",
            "Lights",
            "Smart Devices"
          ],
          "name": "Test-Smart-Home-Skill",
          "summary": "My Personal Smart Home"
        }
      },
      "testingInstructions": "Sample Testing Instructions."
    },
    "permissions": [
      {
        "name": "alexa::async_event:write"
      }
    ],
    "apis": {
      "smartHome": {
        "protocolVersion": "3",
        "endpoint": {
          "uri": "//override"
        }
      }
    }
  }
}