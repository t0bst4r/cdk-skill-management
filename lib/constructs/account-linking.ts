import {Construct, IConstruct} from 'constructs';
import type {ISkill} from './skill';
import {AskCustomResource} from '../ask/ask-custom-resource';
import {SkillAuthenticationProps} from './skill-authentication-props';

/**
 * Represents a platform-specific authorization URL for account linking.
 */
export interface AccountLinkingPlatformAuthorizationUrl {
  /** The platform type */
  readonly platformType: 'iOS' | 'Android';
  /** The platform-specific authorization URL. */
  readonly platformAuthorizationUrl: string;
}

/**
 * Represents the request parameters for account linking.
 */
export interface AccountLinkingRequest {
  /** The type of account linking */
  readonly authenticationFlowType?: 'AUTH_CODE' | 'IMPLICIT';
  /** The authorization URL for account linking. */
  readonly authorizationUrl?: string;
  /** An array of domains for account linking. */
  readonly domains?: Array<string>;
  /** The client ID for account linking. **/
  readonly clientId?: string;
  /** An array of scopes for account linking. */
  readonly scopes?: Array<string>;
  /** The access token URL for account linking. */
  readonly accessTokenUrl?: string;
  /** The client secret for account linking. */
  readonly clientSecret?: string;
  /** The access token scheme. */
  readonly accessTokenScheme?: 'HTTP_BASIC' | 'REQUEST_BODY_CREDENTIALS';
  /** The default token expiration in seconds for account linking. */
  readonly defaultTokenExpirationInSeconds?: number;
  /** The reciprocal access token URL for account linking. */
  readonly reciprocalAccessTokenUrl?: string;
  /** An array of redirect URLs for account linking. */
  readonly redirectUrl?: Array<string>;
  /** An array of platform-specific authorization URLs for account linking. */
  readonly authorizationUrlsByPlatform?: Array<AccountLinkingPlatformAuthorizationUrl>;
  /** Indicates whether to skip account linking on enablement. */
  readonly skipOnEnablement?: boolean;
  /** Voice-forward account linking setting, either 'ENABLED' or 'DISABLED'. */
  readonly voiceForwardAccountLinking?: 'ENABLED' | 'DISABLED';
}

/**
 * Interface representing an Account Linking resource.
 */
export interface IAccountLinking extends IConstruct {}

/**
 * Properties for creating an Account Linking resource.
 */
export interface AccountLinkingProps extends SkillAuthenticationProps {
  /** The Alexa Skill for which account linking is configured. */
  readonly skill: ISkill;
  /** The request parameters for account linking. */
  readonly request: AccountLinkingRequest;
}

/**
 * Represents an Account Linking resource for an Alexa Skill.
 */
export class AccountLinking extends Construct implements IAccountLinking {
  private readonly resource: AskCustomResource;

  /**
   * Creates an instance of the Account Linking resource.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Account Linking properties.
   */
  constructor(scope: Construct, id: string, props: AccountLinkingProps) {
    super(scope, id);

    this.resource = new AskCustomResource(this, 'Default', {
      authenticationConfiguration: props.authenticationConfiguration,
      authenticationConfigurationSecret: props.authenticationConfigurationSecret,
      authenticationConfigurationParameter: props.authenticationConfigurationParameter,
      onUpdate: {
        action: 'updateAccountLinkingInfoV1',
        parameters: [
          props.skill.skillId,
          props.skill.skillStage,
          {
            accountLinkingRequest: {
              ...props.request,
              type: props.request.authenticationFlowType,
              authenticationFlowType: undefined,
            },
          },
        ],
      },
      onDelete: {
        action: 'deleteAccountLinkingInfoV1',
        parameters: [props.skill.skillId, props.skill.skillStage],
      },
    });
    this.resource.node.addDependency(props.skill);
  }
}
