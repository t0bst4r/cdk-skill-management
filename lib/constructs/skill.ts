import {IResource, Resource} from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {CfnSkill} from 'aws-cdk-lib/alexa-ask';
import {Role, ServicePrincipal} from 'aws-cdk-lib/aws-iam';
import {Asset} from 'aws-cdk-lib/aws-s3-assets';
import {SkillAuthenticationProps} from './skill-authentication-props';
import {ISecret} from 'aws-cdk-lib/aws-secretsmanager';
import {IStringParameter} from 'aws-cdk-lib/aws-ssm';

/**
 * Enumeration for different Alexa Skill types.
 */
export enum SkillType {
  CUSTOM = 'CUSTOM',
  SMART_HOME = 'SMART_HOME',
}

/**
 * Interface representing an Alexa Skill.
 */
export interface ISkill extends IResource {
  /** The unique ID of the Alexa Skill. */
  readonly skillId: string;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
}

/**
 * Base class for the Alexa Skill construct.
 */
abstract class SkillBase extends Resource implements ISkill {
  /**
   * The unique ID of the Alexa Skill.
   */
  public abstract readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public abstract readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public abstract readonly skillType: SkillType;
}

/**
 * Type alias for Alexa Skill package overrides.
 */
export type SkillPackageOverrides = CfnSkill.OverridesProperty;

/**
 * Interface representing an Alexa Skill package.
 */
export interface SkillPackage {
  /** The asset associated with the Alexa Skill package. */
  readonly asset: Asset;
  /** Overrides for the Alexa Skill package. */
  readonly overrides?: SkillPackageOverrides;
}

/**
 * Properties for creating an Alexa Skill.
 */
export interface SkillProps extends SkillAuthenticationProps {
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The vendor ID of the Alexa Skill. */
  readonly vendorId: string;
  /** The package information for the Alexa Skill. */
  readonly skillPackage: SkillPackage;
}

/**
 * Properties for creating an Alexa Skill Instance from Attributes.
 */
export interface SkillAttributes {
  /**
   * The unique ID of the Alexa Skill.
   */
  readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  readonly skillType: SkillType;
}

/**
 * Alexa Skill construct for managing an Alexa Skill via CloudFormation.
 */
export class Skill extends SkillBase {
  /**
   * Factory method to create an Alexa Skill from its ID.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skillId - The ID of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromSkillId(scope: Construct, id: string, skillId: string): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = skillId;

      get skillType(): SkillType {
        throw new Error('Access to skillType is not supported when using fromSkillId');
      }

      get skillStage(): string {
        throw new Error('Access to skillStage is not supported when using fromSkillId');
      }
    })(scope, id);
  }

  /**
   * Factory method to create an Alexa Skill from its attributes.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param attributes - The attributes of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromAttributes(scope: Construct, id: string, attributes: SkillAttributes): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = attributes.skillId;
      readonly skillType: SkillType = attributes.skillType;
      readonly skillStage: string = attributes.skillStage;
    })(scope, id);
  }

  private readonly packageRole: Role;
  private readonly resource: CfnSkill;

  /**
   * The unique ID of the Alexa Skill.
   */
  public readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public readonly skillType: SkillType;

  /**
   * Creates an instance of the Alexa Skill construct.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Alexa Skill properties.
   */
  constructor(scope: Construct, id: string, props: SkillProps) {
    super(scope, id);

    const authPropsCount = [
      props.authenticationConfiguration,
      props.authenticationConfigurationParameter,
      props.authenticationConfigurationSecret,
    ].filter(it => !!it).length;

    if (authPropsCount !== 1) {
      throw new Error('Exactly one authentication configuration needs to be provided!');
    }

    this.packageRole = new Role(this, 'PackageRole', {
      assumedBy: new ServicePrincipal('alexa-appkit.amazon.com'),
    });

    const authentication: CfnSkill.AuthenticationConfigurationProperty =
      this.authenticationSecret(props.authenticationConfigurationSecret) ??
      this.authenticationParameter(props.authenticationConfigurationParameter) ??
      props.authenticationConfiguration!;

    this.resource = new CfnSkill(this, 'Default', {
      skillPackage: {
        s3Bucket: props.skillPackage.asset.s3BucketName,
        s3Key: props.skillPackage.asset.s3ObjectKey,
        s3BucketRole: this.packageRole.roleArn,
        overrides: {
          ...props.skillPackage.overrides,
          manifest: {
            ...props.skillPackage.overrides?.manifest,
            publishingInformation: {
              category: props.skillType,
              ...props.skillPackage.overrides?.manifest?.publishingInformation,
            },
          },
        },
      },
      vendorId: props.vendorId,
      authenticationConfiguration: authentication,
    });
    this.resource.node.addDependency(this.packageRole);
    this.resource.node.addDependency(props.skillPackage.asset);
    props.skillPackage.asset.grantRead(this.packageRole);

    this.skillId = this.resource.ref;
    this.skillStage = props.skillStage;
    this.skillType = props.skillType;
  }

  private authenticationSecret(secret: ISecret | undefined): CfnSkill.AuthenticationConfigurationProperty | undefined {
    if (!secret) {
      return;
    }
    return {
      clientId: secret.secretValueFromJson('clientId').unsafeUnwrap(),
      clientSecret: secret.secretValueFromJson('clientSecret').unsafeUnwrap(),
      refreshToken: secret.secretValueFromJson('refreshToken').unsafeUnwrap(),
    };
  }

  private authenticationParameter(
    parameter: IStringParameter | undefined
  ): CfnSkill.AuthenticationConfigurationProperty | undefined {
    if (!parameter) {
      return;
    }
    return JSON.parse(parameter.stringValue);
  }
}
