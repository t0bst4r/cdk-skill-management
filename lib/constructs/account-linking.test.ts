import {App, Arn, ArnFormat, Stack} from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import {Skill, SkillType} from './skill';
import {Secret} from 'aws-cdk-lib/aws-secretsmanager';
import {AccountLinking} from './account-linking';

test('AccountLinking', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  const skillCredentials = Secret.fromSecretCompleteArn(
    stack,
    'SkillCredentials',
    Arn.format(
      {
        arnFormat: ArnFormat.COLON_RESOURCE_NAME,
        service: 'secretsmanager',
        resource: 'secret',
        resourceName: '/Skill/Credentials',
      },
      stack
    )
  );

  const skill = Skill.fromAttributes(stack, 'Skill', {
    skillId: 'My-Skill-Id',
    skillStage: 'development',
    skillType: SkillType.SMART_HOME,
  });
  new AccountLinking(stack, 'AccountLinking', {
    skill,
    request: {
      authenticationFlowType: 'AUTH_CODE',
      authorizationUrl: `https://example.com/auth/authorize`,
      accessTokenUrl: `https://example.com/auth/token`,
      accessTokenScheme: 'REQUEST_BODY_CREDENTIALS',
      clientId: 'client-id',
      clientSecret: 'client-secret',
      scopes: ['smart_home'],
    },
    authenticationConfigurationSecret: skillCredentials,
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});
