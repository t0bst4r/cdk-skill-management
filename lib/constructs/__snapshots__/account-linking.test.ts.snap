// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountLinking 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AccountLinkingC99A8611": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "AccountLinkingProviderRolePolicy866FE269",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomASKCustomResourceProviderHandlerF5F0D5CE",
            "Arn",
          ],
        },
        "authenticationConfigurationSecret": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":secret:/Skill/Credentials",
            ],
          ],
        },
        "delete": "{"action":"deleteAccountLinkingInfoV1","parameters":["My-Skill-Id","development"]}",
        "update": "{"action":"updateAccountLinkingInfoV1","parameters":["My-Skill-Id","development",{"accountLinkingRequest":{"authorizationUrl":"https://example.com/auth/authorize","accessTokenUrl":"https://example.com/auth/token","accessTokenScheme":"REQUEST_BODY_CREDENTIALS","clientId":"client-id","clientSecret":"client-secret","scopes":["smart_home"],"type":"AUTH_CODE"}}]}",
      },
      "Type": "Custom::ASK",
      "UpdateReplacePolicy": "Delete",
    },
    "AccountLinkingProviderRolePolicy866FE269": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":secretsmanager:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:/Skill/Credentials",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AccountLinkingProviderRolePolicy866FE269",
        "Roles": [
          {
            "Fn::Select": [
              1,
              {
                "Fn::Split": [
                  "/",
                  {
                    "Fn::Select": [
                      5,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "CustomASKCustomResourceProviderRoleB79D76EE",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomASKCustomResourceProviderHandlerF5F0D5CE": {
      "DependsOn": [
        "CustomASKCustomResourceProviderRoleB79D76EE",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "13325ce6b32eb5ce62b1c0b20fa16cf263f67b6909b46ee9dea609b97478f866.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomASKCustomResourceProviderRoleB79D76EE",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomASKCustomResourceProviderRoleB79D76EE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
